### PR TITLE
Add mirror support for cask source downloads

### PR DESCRIPTION
**Pull request by Bo98**
See #14690.

### CODE REVIEW

File: Library/Homebrew/env_config.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/env_config.rb:25]: Given folks are mirroring formulae.brew.sh content already: I'd probably just use the existing `HOMEBREW_API_DEFAULT_DOMAIN` as the primary if it's set and fall back to raw rather than having to have yet another format/variable.

**Review by MikeMcQuaid — COMMENTED** [review]
Looks good so far bar the extra variable.

File: Library/Homebrew/env_config.rb
Comment (Bo98) [reply Library/Homebrew/env_config.rb:25]: We don't support cask-source mirroring because it's temporary/private API and I don't think there's an existing formulae.brew.sh mirror in India (but there is for raw).

We were planning to potentially remove cask-source at a later date.

Comment (MikeMcQuaid) [reply Library/Homebrew/env_config.rb:25]: > We ideally want a URL with the commit hash in it, and any mirror should have that as well.

Ours doesn't?

> but there is for raw

Source/URL for this mirror? This feels like a much more appropriate use of a caching proxy rather than a mirror. I've not seen anyone request what's being added here?

**Review by MikeMcQuaid — CHANGES_REQUESTED** [review]
Just to be explicit: let's hold off merging this please until we've had a larger product conversation here.

File: Library/Homebrew/env_config.rb
Comment (Bo98) [reply Library/Homebrew/env_config.rb:25]: > Ours doesn't?

Yes. Which is why it is secondary, only used in emergencies, and may be removed eventually.

> Source/URL for this mirror?

The one provided in that issue is https://www.jsdelivr.com/?docs=gh. I've seen a few others like this that I can probably find and dig out.

Comment (MikeMcQuaid) [reply Library/Homebrew/env_config.rb:25]: Does formulae.brew.sh work for the India folks? If so, why is our fallback not working as expected here? If not, aren't they going to have to mirror that too?

I'm also a bit concerned about adding new mirrors (although this includes the formulae.brew.sh one too that probably should have had more review/thoughts before we merged it) where we unconditionally accept whatever we are sent. If you convince users to set a mirror, you can make them install whatever you want. 

This is less of a concern for bottles but much more of a concern when you deliver formulae or casks that can run pretty much whatever the attacker wants.

Again, as said on Slack, I think this points to us trying to make the API work for use cases where it's not going to be a good fit because of how much trust has to be put into third parties.

Comment (Bo98) [reply Library/Homebrew/env_config.rb:25]: Regarding security: we definitely should be doing better regardless of what this PR does. We'll discuss that elsewhere (and this PR won't progress until after that - don't worry).

> Does formulae.brew.sh work for the India folks? 

I think so

> If so, why is our fallback not working as expected here?

Something with timeout handling I reckon. Not entirely sure what's happening but a potential theory could be that it opens a connection but then the connection gets drops on the remote end without telling the client. In any case the `--speed` stuff will help there too.

To be clear though about the fallback: I would be very much disappointed to see formulae.brew.sh/api/cask-source be relied on (beyond emergencies). It can very easily be desynced with the local formula.json.

**Pull request closed without merging**
